### DISCUSSION TITLE
Downgrade to CSR

### DISCUSSION THREAD

**Discussion by Jerry-Hong** [remix-run/remix#3696] [Proposals]
### What is the new or updated feature that you are suggesting?

Remix can build a `index.html` for all client side render like a SPA.

### Why should this feature be included?

When server is over loading(RenderToString is the bottleneck), we can failover to client side render.

**Comment by sergiodxa** [comment]
Let me see if I understood you, you mean have a way to tell Remix that instead of using the data fetched server side on SSR to render the routes to HTML, it should send a static HTML with the data inlined so it can do the render client-side?

What would happen if your server is over loaded and this happen just when Google is indexing your app? Now your SSR is empty so it will index it that way.

I think if SSR React to an HTML string is the thing slowing you down you probably want to cache the document request, even if that means users will get old data on the first request.

**Reply by sergiodxa** [reply]
BTW, it should be possible to achieve this, the EntryContext object you receive in the entry.server file has all the data needed to render the string, and load the correct JS assets and other link tags.

So you should be able to use that to generate the correct HTML file without the SSR part.

**Reply by Jerry-Hong** [reply]
Really thanks for your reply.

> Let me see if I understood you, you mean have a way to tell Remix that instead of using the data fetched server side on SSR to render the routes to HTML, it should send a static HTML with the data inlined so it can do the render client-side?

Yes, but it dose not need data inlined in HTML just a blank static HTML with scripts like pure SPA first time render.

> What would happen if your server is over loaded and this happen just when Google is indexing your app? Now your SSR is empty so it will index it that way.

If it was a good crawler bot like Google, we will always do SSR.

> I think if SSR React to an HTML string is the thing slowing you down you probably want to cache the document request, even if that means users will get old data on the first request.

In some cases cache is the first option for peak load shifting, but sometime our content cannot show stale content.

> BTW, it should be possible to achieve this, the EntryContext object you receive in the entry.server file has all the data needed to render the string, and load the correct JS assets and other link tags.
>
> So you should be able to use that to generate the correct HTML file without the SSR part.

There is an issue that Remix has not current route match(like `/`) in `window.__remixRouteModules` at first time client side render it will throw error. I need to include all routes to `window.__remixRouteModules` and cannot visit 404 page at first time.
